flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultVertexParallelismAndInputInfosDecider.java

```diff
@@ -126,52 +136,101 @@ public ParallelismAndInputInfos decideParallelismAndInputInfosForVertex(
+                    vertexMaxParallelism,
+                    maxParallelism,
+                    vertexMaxParallelism,
+                    jobVertexId);
+            maxParallelism = vertexMaxParallelism;
+        }
+        checkState(maxParallelism >= minParallelism);
+
+        int parallelism =
+                vertexInitialParallelism > 0
+                        ? vertexInitialParallelism
+                        : decideParallelism(
+                                jobVertexId, consumedResults, minParallelism, maxParallelism);
+
+        Map<Boolean, List<BlockingInputInfo>> inputsGroupByInterCorrelation =
+                consumedResults.stream()
+                        .collect(
+                                Collectors.groupingBy(
+                                        BlockingInputInfo::existInterInputsKeyCorrelation));
+
+        // For AllToAll like inputs, we derive parallelism as a whole, while for Pointwise inputs,
+        // we need to derive parallelism separately for each input.
+        //
+        // In the following cases, we need to reset min parallelism and max parallelism to ensure
+        // that the decide parallelism for all inputs is consistent :
+        // 1.  Vertex has a specified parallelism
+        // 2.  There are edges that don't need to follow intergroup constraint
+        if (vertexInitialParallelism > 0 || inputsGroupByInterCorrelation.containsKey(false)) {
+            minParallelism = parallelism;
+            maxParallelism = parallelism;
+        }
+
+        Map<IntermediateDataSetID, JobVertexInputInfo> vertexInputInfoMap = new HashMap<>();
+
+        if (inputsGroupByInterCorrelation.containsKey(true)) {
+            vertexInputInfoMap.putAll(
+                    allToAllVertexInputInfoComputer.compute(
+                            jobVertexId,
+                            inputsGroupByInterCorrelation.get(true),
+                            parallelism,
+                            minParallelism,
+                            maxParallelism));
+        }
+
+        if (inputsGroupByInterCorrelation.containsKey(false)) {
+            List<BlockingInputInfo> inputsWithoutInterCorrelation =
+                    inputsGroupByInterCorrelation.get(false);
+            for (BlockingInputInfo input : inputsWithoutInterCorrelation) {
+                if (input.existIntraInputKeyCorrelation()) {
+                    vertexInputInfoMap.putAll(
+                            allToAllVertexInputInfoComputer.compute(
+                                    jobVertexId,
+                                    Collections.singletonList(input),
+                                    parallelism,
+                                    minParallelism,
+                                    maxParallelism));
+                } else {
+                    vertexInputInfoMap.put(
+                            input.getResultId(),
+                            pointwiseVertexInputInfoComputer.compute(input, parallelism));
```

Comment:
Why not existing `existIntraInputKeyCorrelation` leads to `pointwiseVertexInputInfoComputer#compute()`?
`existIntraInputKeyCorrelation` is false for `rebalance` cases (all-to-all)?